src/transform/tests/testdata/union-fusion

```diff
@@ -0,0 +1,57 @@
+# Copyright Materialize, Inc. and contributors. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+cat
+(defsource x [int64 int64])
+----
+ok
+
+build apply=UnionFusion
+(union [(get x) (union [(get x) (union [(get x) (get x)])])])
```

Comment:
sorry for the drive by review but this piqued my interest. Is this optimization dependent on the unions appearing in the order of this test, and if yes can real world queries end up like this?:

```
 (union [(union [(union [(get x) (get x)]) (get x)]) (get x)])
```

Reply:
No worries. Feedback is always appreciated!

The optimization doesn't depend on the order of the branches. Any union branch that is a union (or a negated union) will be fused into the parent union by this transform.

The order in your example is perfectly reproducible through SQL by placing parenthesis to shape the union tree that way:

```
materialize=> explain raw plan for select * from t1 union all select * from t1 union all select * from t1 union all select * from t1;;
             Raw Plan              
-----------------------------------
 %0 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %1 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %2 =                             +
 | Union %0 %1                    +
                                  +
 %3 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %4 =                             +
 | Union %2 %3                    +
                                  +
 %5 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %6 =                             +
 | Union %4 %5                    +
 
(1 row)

materialize=> explain raw plan for select * from t1 union all (select * from t1 union all (select * from t1 union all select * from t1));
             Raw Plan              
-----------------------------------
 %0 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %1 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %2 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %3 =                             +
 | Get materialize.public.t1 (u43)+
                                  +
 %4 =                             +
 | Union %2 %3                    +
                                  +
 %5 =                             +
 | Union %1 %4                    +
                                  +
 %6 =                             +
 | Union %0 %5                    +
 
(1 row)

materialize=> 
```